File: 灵神/basicPractice/Java/src/com/kurumi/basicPractice/动态规划/LeetCode_0746.java
package com.kurumi.basicPractice.动态规划;

/**
 * @Author kurumi
 * @Date 2026-07-08 下午4:41:59
 * @Description 746. 使用最小花费爬楼梯
 */
public class LeetCode_0746 {
    static class Solution {
        public int minCostClimbingStairs(int[] cost) {
            int n = cost.length;
            int[] dp = new int[n + 1];
            for (int i = 2; i <= n; i++) dp[i] = Math.min(dp[i - 1] + cost[i - 1], dp[i - 2] + cost[i - 2]);
            return dp[n];
        }
    }

    static class Solution1 {
        public int minCostClimbingStairs(int[] cost) {
            int n = cost.length;
            int d0 = 0, d1 = 0;
            for (int i = 2; i <= n; i++) {
                int preD1 = d1;
                d1 = Math.min(d1 + cost[i - 1], d0 + cost[i - 2]);
                d0 = preD1;
            }
            return d1;
        }
    }
}

File: 灵神/basicPractice/Python/动态规划/LeetCode_0746.py
'''
@Author kurumi
@Date 2026-07-08 下午4:41:59
@Description 746. 使用最小花费爬楼梯
'''
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n: int = len(cost)
        dp: list = [0] * (n + 1)
        for i in range(2, n + 1):
            dp[i] = min(dp[i - 1] + cost[i - 1], dp[i - 2] + cost[i - 2])
        return dp[n]

class Solution1:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n: int = len(cost)
        d0, d1 = 0, 0
        for i in range(2, n + 1):
            d0, d1 = d1, min(d1 + cost[i - 1], d0 + cost[i - 2])
        return d1